### ISSUE TITLE
Figure out how to use the OJB device

### ISSUE THREAD

**Issue by eswenson1** [PDP-10/its#1468]
Turns out there is an OJB device (in addition to the JOB and BOJ devices).  It *appears* that the difference between the JOB device and the OJB device is that whereas JOB launches a job to satisfy the request, OJB searches for an existing user job to satisfy the request. This may well be a way to debug JOB handlers (e.g. DEVICE; ATSIGN MLDEV).  My theory is that you could start up an MLDEV (normal user job) under DDT and then use OBJ: rather than JOB: to use it as the handler for an I/O request to device.  

SYSDOC; JOB 100 describes the JOB and BOJ devices, but there is no mention of the OBJ device.  I did find references to it in SYSTEM; ITS 1648 and the comments in ITS suggest that it behaves as mentioned above.

**Comment by larsbrinkhoff** [comment]
The above says both OBJ and OJB.  The latter is correct, right?

**Comment by eswenson1** [comment]
Right. OJB is correct. I can’t ever type that correctly!

**Comment by larsbrinkhoff** [comment]
I see OPTOBJ is defined, but the symbol is never used.  At least it doesn't show up in searches.

**Comment by larsbrinkhoff** [comment]
Oh sorry, it's %OPOBJ.

**Comment by eswenson1** [comment]
I haven't been able to get this to work.  I created a job (MLDEV) in one HACTRN, logged in as EJS and loaded device;atsign mldev.  In another HACTRN (logged in as EJS0), I created a translation between JOB:device;atsign mldev and OJB: EJS; EJS MLDEV.  I then tried to do a ^R NO:lars; and just got:

```
NO: LARS; EJS MLDEV - DEVICE NOT AVAILABLE
```.

However, if I do a ^R OJB:EJS;EJS MLDEV, I return from the BOJ open call in the BDH -- so that definitely does work.  However, I haven't been able to find a translation that works so that doing a ^R no;lars; invokes my MLDEV BDH (under DDT).

I've even tried a translation of JOB:*;* * -> OJB:EJS; EJS MLDEV.  I know that target works (see above), but I can't get an ^R NO;LARS;, which should cause JOB:DEVICE;JOBDEV NO to get invoked, to work.  I've tried the following translations:

```
DB ITS 1648  Peek 630   1/8/2019 15:40:22  Up time = 4:16:43:37
EJS0 HACTRN & inferiors:
   IO   JOB:DEVICE;ATSIGN MLDEV =>  OBJ:EJS;EJS MLDEV
   IO   JOB: *; *  * =>  OJB:EJS;EJS MLDEV
   IO   JOB:DEVICE;JOBDEV NO =>  OJB:EJS;EJS MLDEV
```

**Comment by eswenson1** [comment]
It may well be that what I'm trying to do is impossible. That what I need to do is write a test program that explicitly opens the OJB:EJS;EJS MLDEV device.  And presumably, it will have to support getting invoked from a JOBCAL (from the MLDEV BDH) so that my test program can pass it the dev (NO), dir (lars), FN1 (arbitrary), FN2 (arbitrary) parameters.

I was hoping that when ITS invoked JOB:DEVICE;JOBDEV NO, that it would respect translations, but thinking about it, the translations set with $^T are for DDT and its inferiors.  The invoking program, in this case, is ITS and it probably doesn't use the translations to invoke the OBJ: device rather than the JOB: device.  I'd have to run a program under DDT (that was subject to translations) in order to do that.

This is supported by the fact that an explicit ^R OJB:EJS;EJS MLDEV hangs (assuming the BDH is stopped at a breakpoint) rather than giving me a DEVICE NOT AVAILABLE error.  In other words, the explicit reference to the OJB device does invoke the BDH under DDT.

**Comment by eswenson1** [comment]
I think I got this to work with the MLDEV device.  In one HACTRN (EJS), I created a job (MLDEV$J).  In another HACTRN (EJS0), I set up a $^T translation for NO:,OBJ:EJS;EJS MLDEV.  I then, in that HACTRN, did a ^R NO:LARS;  This hung (as expected).  I went back to the EJS HACTRN, and then did a $L DEVICE; ATSIGN MLDEV.  I started stepping from the start address, and the JOBCAL, succeeded.  When I looked at JBCFN1 I found .FILE., in JBCFN2 I found (DIR), in JBCSNM, I found LARS and in JBCDEV, I found NO.  This is as expected.  I was then able to debug MLDEV and it was handling the request as through I had done a ^R NO:LARS; (in other words, listing the LARS directory on the NO device (machine).  So it *does* work fine.  When I ran the MLDEV BDH, and switched back to my EJS0 HACTRN, I saw the output of the LARS directory on NO.  Next, I'm going to try this with ARCDEV.

**Comment by eswenson1** [comment]
Experiment worked equally well with ARCDEV.  My translation was AR1:,OJB:EJS;EJS ARCDEV and my BDH was (obviously) ARCDEV.

**Comment by eswenson1** [comment]
So I've successfully debugged ARCDEV 69 and replicated the "issue" we've seen with ARCDEV 68 and 69.  I've stepped through the OPEN and write calls (which were, I think, successful).  However, the BDH (ARCDEV) is getting an interrupt (loc 42) at some point -- and I didn't debug enough to know exactly at what point.  The interrupt is an IOC interrupt (not sure what this is, yet).  This causes ARCDEV to call the subroutine TSIOC.  The comment above this routine says:

```
;IOC error must be disk full.  Just wait a while and try again.
;(could also be directory full, but what can we do?)
```

I suspect the interpretation of the interrupt is incorrect because there is no disk / directory full condition.  The code just sleeps and tries again.  We get the interrupt again and again and this goes on ad infinitum.  ARCDEV never continues to completion and termination.  I'll have to debug further to figure out what the interrupt is really for.

**Comment by larsbrinkhoff** [comment]
So is debugging with OJB figured out now?

**Comment by eswenson1** [comment]
Yes. I should write up an example. Where? We should have a tips and tricks doc file. Suggestions for location?

**Comment by larsbrinkhoff** [comment]
If it's not appropriate to add to SYSDOC; JOB 100, then some other
file in SYSDOC.

**Comment by eswenson1** [comment]
There is already some doc in JOB 100. However, I was thinking more of a concrete example. Also, I was thinking of a new file, something like _INFO_; TIPS >, where we could have a random assortmemt of tips that we could collect things we’ve figured out and that others might find helpful.

**Comment by eswenson1** [comment]
I decided to take the following approach: update info;dir > to include a new menu item (Tips).  Have it reference a new file: info;tips >.  Use INFO to add topics of interest.  We can add to them and organize them as we see fit.  I've added one topic (locally) about how to build a bootable ITS (e.g. create .;@ its).  Will add another about nsalv/salv.  And another about how to debug a BDH.  And another about adding packs, etc.
